rs/memory_tracker/src/lib.rs

```diff
@@ -126,69 +126,113 @@ impl PageBitmap {
+        let mut end = faulting_page.get();
         while end < old_end {
-            if self.pages.get(end).unwrap_or(true) {
+            if self.pages.get(end as usize).unwrap_or(true) {
                 break;
             }
             end += 1;
         }
         Range {
-            start: PageIndex::new(start as u64),
-            end: PageIndex::new(end as u64),
+            start: PageIndex::new(start),
+            end: PageIndex::new(end),
         }
     }
 
-    // Returns the largest prefix of the given range such that all pages there have
-    // been already marked.
-    fn restrict_range_to_marked(&self, range: Range<PageIndex>) -> Range<PageIndex> {
+    // Returns the largest range around the faulting page such that
+    // all pages there have been already marked.
+    fn restrict_range_to_marked(
+        &self,
+        faulting_page: PageIndex,
+        range: Range<PageIndex>,
+    ) -> Range<PageIndex> {
         let range = range_intersection(&range, &self.page_range());
 
-        let start = range.start.get() as usize;
-        let old_end = range.end.get() as usize;
-        let mut end = start;
+        let old_start = range.start.get();
+        let mut start = faulting_page.get();
+        while start > old_start {
+            match self.pages.get(start as usize - 1) {
+                None | Some(false) => break,
+                Some(true) => {}
+            }
+            start -= 1;
+        }
+        let old_end = range.end.get();
+        let mut end = faulting_page.get();
         while end < old_end {
-            match self.pages.get(end) {
+            match self.pages.get(end as usize) {
                 None | Some(false) => break,
                 Some(true) => {}
             }
             end += 1;
         }
         Range {
-            start: PageIndex::new(start as u64),
-            end: PageIndex::new(end as u64),
+            start: PageIndex::new(start),
+            end: PageIndex::new(end),
         }
     }
 
     // Returns the range of pages that are predicted to be marked in the future
-    // based on the marked pages before the start of the given range.
-    fn restrict_range_to_predicted(&self, range: Range<PageIndex>) -> Range<PageIndex> {
+    // based on the marked pages before the start of the given range or after the end.
+    fn restrict_range_to_predicted(
+        &self,
+        faulting_page: PageIndex,
+        range: Range<PageIndex>,
+    ) -> Range<PageIndex> {
         let range = range_intersection(&range, &self.page_range());
         if range.is_empty() {
             return range;
         }
 
-        let start = range.start.get() as usize;
-        let old_end = range.end.get() as usize;
+        let page = faulting_page.get();
+        let start = range.start.get();
+        let end = range.end.get();
+
+        let mut bwd_predicted_count = 0;
+        while page - bwd_predicted_count > start && bwd_predicted_count < page {
```

Comment:
Isn't `bwd_predicted_count < page` implied by `page - bwd_predicted_count > start` because these are all `PageIndex` which is non-negative? Or maybe we should be checking `bwd_predicted_count <= page` first to ensure the subtraction doesn't trap?

Reply:
Both `page` and `bwd_predicted_count` are `u64`s here, so the check is redundant.